File: golang/559_max_depth/max_depth.go
/*
 * @Date: 2021-11-21 02:11:21
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-11-27 00:35:02
 */

package main

type Node struct {
	Val      int
	Children []*Node
}

func maxDepth(root *Node) (ans int) {
	if root == nil {
		return
	}
	queue := []*Node{root}
	for len(queue) > 0 {
		q := queue
		queue = nil
		for _, node := range q {
			queue = append(queue, node.Children...)
		}
		ans++
	}
	return
}

func maxDepth2(root *Node) int {
	if root == nil {
		return 0
	}
	maxChildDepth := 0
	for _, child := range root.Children {
		if childDepth := maxDepth(child); childDepth > maxChildDepth {
			maxChildDepth = childDepth
		}
	}
	return maxChildDepth + 1
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}

	{
		root := &Node{1, []*Node{{2, nil}, {3, []*Node{{5, nil}, {6, nil}}}, {4, nil}}}
		assert(maxDepth(root) == 3)
	}
	{
		root := &Node{1, []*Node{{2, nil}, {3, []*Node{{6, nil}, {7, []*Node{{11, []*Node{{12, nil}}}}}}}, {4, []*Node{{8, []*Node{{12, nil}}}}}, {5, []*Node{{9, nil}, {10, nil}}}}}
		assert(maxDepth(root) == 5)
	}
}

File: rust/1614_max_depth/max_depth.rs
/*
 * @Date: 2022-01-07 01:08:47
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-01-07 01:25:02
 */

pub fn max_depth(s: String) -> i32 {
    s.chars()
        .fold((0, 0), |(mx, cnt), c| match c {
            '(' => (mx.max(cnt + 1), cnt + 1),
            ')' => (mx, cnt - 1),
            _ => (mx, cnt),
        })
        .0
}

fn main() {
    assert_eq!(max_depth("(1+(2*3)+((8)/4))+1".to_string()), 3);
    assert_eq!(max_depth("(1)+((2))+(((3)))".to_string()), 3);
    assert_eq!(max_depth("1+(2*3)/(2-1)".to_string()), 1);
    assert_eq!(max_depth("1".to_string()), 0);
}
